dtfs-central-api/src/services/state-machines/fee-record/event-handlers/keying-data-generated/keying-data-generated.event-handler.ts

```diff
@@ -0,0 +1,23 @@
+import { EntityManager } from 'typeorm';
+import { DbRequestSource, FeeRecordEntity } from '@ukef/dtfs2-common';
+import { BaseFeeRecordEvent } from '../../event/base-fee-record.event';
+
+type KeyingDataGeneratedEventPayload = {
+  transactionEntityManager: EntityManager;
+  generateKeyingData: boolean;
+  requestSource: DbRequestSource;
+};
+
+export type FeeRecordKeyingDataGeneratedEvent = BaseFeeRecordEvent<'KEYING_DATA_GENERATED', KeyingDataGeneratedEventPayload>;
+
+export const handleFeeRecordKeyingDataGeneratedEvent = async (
+  feeRecord: FeeRecordEntity,
+  { transactionEntityManager, generateKeyingData, requestSource }: KeyingDataGeneratedEventPayload,
+): Promise<FeeRecordEntity> => {
```

Comment:
I wonder if this should actually be named `GENERATE_KEYING_DATA` since it is this handler that does that actual generation, it's not just some side-effect status update like `PAYMENT_DELETED` is? That does make it a bit tricky for the payload property, `generateKeyingData` but I almost wonder if that parameter should actually just be named `isLastFeeRecordForFacility` and then this function owns knowing that being the last fee record for a facility or not corresponds to whether we calculate those specific fields or not?

Reply:
Whilst the computation is complex, the concept is simple, so I do think `isLastFeeRecordForFacility` will work as it is just renaming a parameter no?

Reply:
Discussed offline: to remove any implication about ordering but make the intention more clear, we have decided to go with `isFinalFeeRecordForFacility`